### PR TITLE
NewExhaustiveProgram to evaluate without short-circuiting

### PR DESCRIPTION
**Pull request by jiyunyao**
Standard evaluation will short-circuit conditionals and certain macro comprehensions,
but with this is not always desired. When debugging why a certain expression evaluates
the way it does, the short-cirtcuit is an obstacle to iteratively overcome. This feature
disables short-circuits for the majority of the code paths, but does not yet cover comprehension
macros.

This PR enables #121 for all non-comprehension paths.

### CODE REVIEW

File: interpreter/astwalker.go [hunk 1]
```diff
@@ -152,11 +155,13 @@ func (w *astWalker) walkCall(node *exprpb.Expr) []Instruction {
 			evalCount += argGroupLens[i]
 			instructions = append(instructions, argGroup...)
 			if i != argCount-1 {
-				instructions = append(instructions,
-					NewJump(
-						argIDs[i],
-						instructionCount-evalCount,
-						jumpIfEqual(argIDs[i], types.Bool(function == operators.LogicalOr))))
+				if w.forceEval{
```
Comment (TristonianJones) [inline interpreter/astwalker.go]: Here and below, please consider phrasing this as though evaluating all branches is disabled by default.
```
if i != argCount - 1 && w.shortcircuit {
  instructions = append(instructions, NewJump(...))
}
```

File: interpreter/astwalker.go [hunk 2]
```diff
@@ -38,15 +38,17 @@ const (
 func WalkExpr(expression *exprpb.Expr,
 	metadata Metadata,
 	dispatcher Dispatcher,
-	state MutableEvalState) []Instruction {
+	state MutableEvalState,
+	forceEval bool) []Instruction {
```
Comment (TristonianJones) [inline interpreter/astwalker.go]: Consider renaming from `forceEval` to `shortcircuit`.  Removing the `jump` instructions is all that's needed to evaluate all branches.

File: interpreter/astwalker.go [hunk 3]
```diff
@@ -185,14 +190,22 @@ func (w *astWalker) walkCall(node *exprpb.Expr) []Instruction {
 			NewJump(conditionID, len(trueVal)+len(falseVal)+3,
 				jumpIfUnknownOrError(conditionID)))
 		// 2: jump to <ELSE> on false.
-		instructions = append(instructions,
-			NewJump(conditionID, len(trueVal)+2,
-				jumpIfEqual(conditionID, types.False)))
+		if w.forceEval {
```
Comment (TristonianJones) [inline interpreter/astwalker.go]:
```
if w.shortcircuit {
   instructions = append(instructions, NewJump(....))
}
```

File: interpreter/instructions.go [hunk 4]
```diff
@@ -197,6 +197,21 @@ func NewJump(exprID int64, instructionCount int, cond func(EvalState) bool) *Jum
 		OnCondition:     cond}
 }
 
+// NoOp represents a no operational placeholder.
```
Comment (TristonianJones) [inline interpreter/instructions.go]: Per the earlier comments, I'm not sure you need the `NoOp`. Give it a try without them and if it works, then remove this code, and if it doesn't please just rephrase the code as though short-circuiting is enabled by default, e.g.
```
inst := NewJump(...)
if !w.shortcircuit {
  inst := NewNoOp()
}
instructions = append(instructions, inst)
```

File: interpreter/interpreter_test.go [hunk 5]
```diff
@@ -30,6 +30,51 @@ import (
...
+	program := NewCompleteProgram(
+		test.Conditional.Expr,
+		test.Conditional.Info(t.Name()))
+
+	interpretable := interpreter.NewInterpretable(program)
+	result, state := interpretable.Eval(
+		NewActivation(map[string]interface{}{
+			"a": true,
+			"b": 0.999,
+			"c": types.NewStringList([]string{"hello"})}))
```
Comment (TristonianJones) [inline interpreter/interpreter_test.go:46]: Would you mind adding a comment about why the test for the value of `c` is being done?

File: interpreter/interpreter_test.go [hunk 6]
```diff
@@ -30,6 +30,51 @@ import (
...
+	ev, _ := state.Value(6)
```
Comment (TristonianJones) [inline interpreter/interpreter_test.go:47]: Please comment that the identifier for `c` has the expression id `6`.  Same for the later tests where the magic expression identifier is used. You'll want to mention that the next developer should see the test/expr code to figure out what this value should be (or provide a convenient function that gets the expression id for `c` by name so the next person doesn't have to guess).

File: interpreter/program.go [hunk 7]
```diff
@@ -83,7 +84,22 @@ func NewProgram(expression *exprpb.Expr,
...
 		revInstructions: revInstructions,
-		metadata:        newExprMetadata(info)}
+		metadata:        newExprMetadata(info),
+		forceEval:       false,
+	}
+}
+
+// NewCompleteProgram creates a Program from a CEL expression and source
+// information which force evaluating all branches of the expression.
+func NewCompleteProgram(expression *exprpb.Expr,
```
Comment (TristonianJones) [inline interpreter/program.go]: Maybe call this `NewExhaustiveProgram`. The word `Complete` connotes that somehow the regular program is incomplete.

File: interpreter/astwalker.go (hunk 1 above)
Comment (jiyunyao) [reply interpreter/astwalker.go]: Done. Works better~ Thanks

File: interpreter/astwalker.go (hunk 2 above)
Comment (jiyunyao) [reply interpreter/astwalker.go]: Done~

File: interpreter/astwalker.go (hunk 3 above)
Comment (jiyunyao) [reply interpreter/astwalker.go]: Done. Works better~ Thanks

File: interpreter/instructions.go (hunk 4 above)
Comment (jiyunyao) [reply interpreter/instructions.go]: Done. NoOp is removed

File: interpreter/interpreter_test.go (hunk 5 above)
Comment (jiyunyao) [reply interpreter/interpreter_test.go:46]: Comments added~ Please let me know if it is sufficient~

File: interpreter/interpreter_test.go (hunk 6 above)
Comment (jiyunyao) [reply interpreter/interpreter_test.go:47]: Comments added~ Please let me know if it is sufficient~

File: interpreter/program.go (hunk 7 above)
Comment (jiyunyao) [reply interpreter/program.go]: Done.

File: interpreter/interpreter_test.go [hunk 8]
```diff
@@ -155,15 +238,18 @@ func TestInterpreter_LogicalOrEquals(t *testing.T) {
 	provider := types.NewProvider(&exprpb.Expr{})
 	i := NewStandardInterpreter(packages.NewPackage("test"), provider)
 	interpretable := i.NewInterpretable(program)
-	result, _ := interpretable.Eval(
+	result, state := interpretable.Eval(
 		NewActivation(map[string]interface{}{
 			"a": true,
 			"b": "b",
 		}))
+	rhv, _ := state.Value(4)
```
Comment (TristonianJones) [inline interpreter/interpreter_test.go:249]: Please add a comment about where the numeric indices provided to `Value()` calls come from (here and above)

**Review by TristonianJones — CHANGES_REQUESTED** [review]
Looks generally pretty good, but please add a couple of comments to the tests.

File: interpreter/interpreter_test.go (hunk 8 above)
Comment (jiyunyao) [reply interpreter/interpreter_test.go:249]: done

**Comment by jiyunyao** [comment]
Done.

**Review by TristonianJones — COMMENTED** [review]
The code present is in good shape, and it allows for exhaustive evaluation over many of the branches
of execution within a CEL expression, but not all. 

For comprehensions, you will need to change comprehension loop to determine the state at each iteration of the loop. The `EvalState` object doesn't capture multiple evaluations of the same
conditions, so the change is more extensive than what is presented here. 

Please create an Issue as follow:

> Introduce an exhaustive evaluation mode
>
> Exhaustive evaluation does not short-circuit conditions, and iterates through all elements
> in the `iterRange` argument to a comprehension. The state of each iteration should also be
> tracked and returned with the call result.

Once the issue is created and a TODO added to the `NewExhaustiveProgram` call, I'll approve
this change and reference the parent issue on merge.

This is great first step. :+1:

**Comment by jiyunyao** [comment]
#121

**Comment by jiyunyao** [comment]
Done.

**Pull request merged**